File: rln/ffi_c_examples/stateless.c
#include <stdio.h>
#include <stdlib.h>

#include "common.c"

int main(void)
{
    FFI_RLNV3_t *rln_instance = init_rln_stateless();
    if (!rln_instance)
    {
        return EXIT_FAILURE;
    }

    Member_t member;
    create_member(&member);

    printf("\nComputing Merkle path for stateless mode\n");
    CFr_t *default_leaf = ffi_cfr_zero();
    CFr_t *default_hashes[TREE_DEPTH - 1];
    default_hashes[0] = ffi_poseidon_hash_pair(default_leaf, default_leaf);
    for (size_t i = 1; i < TREE_DEPTH - 1; i++)
    {
        default_hashes[i] = ffi_poseidon_hash_pair(default_hashes[i - 1], default_hashes[i - 1]);
    }
    Vec_CFr_t path_elements = ffi_vec_cfr_new(TREE_DEPTH);
    ffi_vec_cfr_push(&path_elements, default_leaf);
    for (size_t i = 1; i < TREE_DEPTH; i++)
    {
        ffi_vec_cfr_push(&path_elements, default_hashes[i - 1]);
    }
    uint8_t path_index_data[TREE_DEPTH] = {0};
    Vec_uint8_t path_index = {path_index_data, TREE_DEPTH, TREE_DEPTH};

    printf("\nComputing Merkle root for stateless mode\n");
    printf("  - computing root for index 0 with rate commitment\n");
    CFr_t *computed_root = ffi_poseidon_hash_pair(member.rate_commitment, default_leaf);
    for (size_t i = 1; i < TREE_DEPTH; i++)
    {
        CFr_t *next_root = ffi_poseidon_hash_pair(computed_root, default_hashes[i - 1]);
        ffi_cfr_free(computed_root);
        computed_root = next_root;
    }
    print_cfr("computed root", computed_root);
    Vec_CFr_t roots = ffi_vec_cfr_new(1);
    ffi_vec_cfr_push(&roots, computed_root);

    CFr_t *external_nullifier = compute_external_nullifier();

    printf("\nHashing signal\n");
    uint8_t signal[32] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 0, 0, 0, 0, 0, 0,
                          0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
    CFr_t *x = hash_signal(signal);
    print_cfr("x", x);

    printf("\nCreating message id\n");
    CFr_t *message_id = ffi_uint_to_cfr(0);
    print_cfr("message id", message_id);

    printf("\nCreating RLN witness\n");
    CResult_FFI_RLNV3WitnessInput_ptr_Vec_uint8_t witness_result =
        ffi_rln_v3_witness_input_new_single(member.identity_secret, member.user_message_limit,
                                            message_id, &path_elements, &path_index, x,
                                            external_nullifier);
    if (!witness_result.ok)
    {
        fprintf(stderr, "Witness creation error: %s\n", witness_result.err.ptr);
        ffi_c_string_free(witness_result.err);
        return EXIT_FAILURE;
    }
    FFI_RLNV3WitnessInput_t *witness = witness_result.ok;
    printf("  - RLN witness created successfully\n");

    printf("\nGenerating RLN proof\n");
    CResult_FFI_RLNV3Proof_ptr_Vec_uint8_t rln_proof_result =
        ffi_rln_v3_generate_proof(&rln_instance, &witness);
    if (!rln_proof_result.ok)
    {
        fprintf(stderr, "Proof generation error: %s\n", rln_proof_result.err.ptr);
        ffi_c_string_free(rln_proof_result.err);
        return EXIT_FAILURE;
    }
    FFI_RLNV3Proof_t *rln_proof = rln_proof_result.ok;
    printf("  - proof generated successfully\n");

    printf("\nGetting RLN proof values\n");
    FFI_RLNV3ProofValues_t *proof_values = ffi_rln_v3_proof_get_values(&rln_proof);
    CResult_CFr_ptr_Vec_uint8_t y_result = ffi_rln_v3_proof_values_get_y(&proof_values);
    if (!y_result.ok)
    {
        fprintf(stderr, "Get y error: %s\n", y_result.err.ptr);
        ffi_c_string_free(y_result.err);
        return EXIT_FAILURE;
    }
    print_cfr("y", y_result.ok);
    ffi_cfr_free(y_result.ok);
    CResult_CFr_ptr_Vec_uint8_t nullifier_result = ffi_rln_v3_proof_values_get_nullifier(&proof_values);
    if (!nullifier_result.ok)
    {
        fprintf(stderr, "Get nullifier error: %s\n", nullifier_result.err.ptr);
        ffi_c_string_free(nullifier_result.err);
        return EXIT_FAILURE;
    }
    print_cfr("nullifier", nullifier_result.ok);
    ffi_cfr_free(nullifier_result.ok);
    CFr_t *proof_values_root = ffi_rln_v3_proof_values_get_root(&proof_values);
    print_cfr("root", proof_values_root);
    ffi_cfr_free(proof_values_root);
    CFr_t *proof_values_x = ffi_rln_v3_proof_values_get_x(&proof_values);
    print_cfr("x", proof_values_x);
    ffi_cfr_free(proof_values_x);
    CFr_t *proof_values_external_nullifier =
        ffi_rln_v3_proof_values_get_external_nullifier(&proof_values);
    print_cfr("external nullifier", proof_values_external_nullifier);
    ffi_cfr_free(proof_values_external_nullifier);

    printf("\nVerifying proof\n");
    CBoolResult_t verify_result =
        ffi_rln_v3_verify_with_roots(&rln_instance, &rln_proof, &roots, x);
    if (verify_result.err.ptr)
    {
        fprintf(stderr, "Proof verification error: %s\n", verify_result.err.ptr);
        ffi_c_string_free(verify_result.err);
        return EXIT_FAILURE;
    }
    if (verify_result.ok)
    {
        printf("  - proof verified successfully\n");
    }
    else
    {
        printf("Proof verification failed\n");
        return EXIT_FAILURE;
    }

    ffi_rln_v3_proof_values_free(proof_values);
    ffi_rln_v3_proof_free(rln_proof);
    ffi_rln_v3_witness_input_free(witness);
    ffi_cfr_free(message_id);
    ffi_cfr_free(x);
    ffi_cfr_free(external_nullifier);
    ffi_vec_cfr_free(roots);
    ffi_cfr_free(computed_root);
    ffi_vec_cfr_free(path_elements);
    for (size_t i = 0; i < TREE_DEPTH - 1; i++)
    {
        ffi_cfr_free(default_hashes[i]);
    }
    ffi_cfr_free(default_leaf);
    member_free(&member);
    ffi_rln_v3_free(rln_instance);
    return EXIT_SUCCESS;
}

File: rln/ffi_nim_examples/stateless.nim
include "common"

proc main() =
  var rlnInstance = initRLNStateless()
  if rlnInstance.isNil:
    return

  var member = createMember()

  echo "\nComputing Merkle path for stateless mode"
  let defaultLeaf = ffi_fr_zero()
  var defaultHashes: array[treeDepth - 1, ptr Fr]
  defaultHashes[0] = ffi_poseidon_hash_pair(defaultLeaf, defaultLeaf)
  for i in 1 ..< treeDepth - 1:
    defaultHashes[i] = ffi_poseidon_hash_pair(defaultHashes[i - 1],
        defaultHashes[i - 1])
  var pathElements = ffi_vec_fr_new(csize_t(treeDepth))
  ffi_vec_fr_push(addr pathElements, defaultLeaf)
  for i in 1 ..< treeDepth:
    ffi_vec_fr_push(addr pathElements, defaultHashes[i - 1])
  var pathIndexData: array[treeDepth, uint8]
  var pathIndex = Vec_uint8(dataPtr: addr pathIndexData[0],
      len: csize_t(treeDepth), cap: csize_t(treeDepth))

  echo "\nComputing Merkle root for stateless mode"
  echo "  - computing root for index 0 with rate commitment"
  var computedRoot = ffi_poseidon_hash_pair(member.rateCommitment, defaultLeaf)
  for i in 1 ..< treeDepth:
    let nextRoot = ffi_poseidon_hash_pair(computedRoot, defaultHashes[i - 1])
    ffi_fr_free(computedRoot)
    computedRoot = nextRoot
  printFr("computed root", computedRoot)
  var roots = ffi_vec_fr_new(csize_t(1))
  ffi_vec_fr_push(addr roots, computedRoot)

  let externalNullifier = computeExternalNullifier()

  echo "\nHashing signal"
  var signal: array[32, uint8] = [1'u8, 2, 3, 4, 5, 6, 7, 8, 9, 10, 0, 0, 0, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
  let x = hashSignal(signal)
  printFr("x", x)

  echo "\nCreating message id"
  let messageId = ffi_uint_to_fr(0'u32)
  printFr("message id", messageId)

  echo "\nCreating RLN witness"
  let merkleProof = ffi_rln_merkle_proof_new(addr pathElements, addr pathIndex)
  let witnessResult = ffi_rln_witness_input_new_single(member.identitySecret,
      member.userMessageLimit, messageId, merkleProof, x, externalNullifier)
  if witnessResult.ok.isNil:
    stderr.writeLine("Witness creation error: " & asString(witnessResult.err))
    ffi_c_string_free(witnessResult.err)
    return
  var witness = witnessResult.ok
  echo "  - RLN witness created successfully"

  echo "\nGenerating RLN proof"
  let rlnProofResult = ffi_rln_generate_proof(rlnInstance, witness)
  if rlnProofResult.ok.isNil:
    stderr.writeLine("Proof generation error: " & asString(rlnProofResult.err))
    ffi_c_string_free(rlnProofResult.err)
    return
  var rlnProof = rlnProofResult.ok
  echo "  - proof generated successfully"

  echo "\nGetting RLN proof values"
  var proofValues = ffi_rln_proof_get_values(rlnProof)
  let yResult = ffi_rln_proof_values_get_y(proofValues)
  if yResult.ok.isNil:
    stderr.writeLine("Get y error: " & asString(yResult.err))
    ffi_c_string_free(yResult.err)
    return
  printFr("y", yResult.ok)
  ffi_fr_free(yResult.ok)
  let nullifierResult = ffi_rln_proof_values_get_nullifier(proofValues)
  if nullifierResult.ok.isNil:
    stderr.writeLine("Get nullifier error: " & asString(nullifierResult.err))
    ffi_c_string_free(nullifierResult.err)
    return
  printFr("nullifier", nullifierResult.ok)
  ffi_fr_free(nullifierResult.ok)
  let proofValuesRoot = ffi_rln_proof_values_get_root(proofValues)
  printFr("root", proofValuesRoot)
  ffi_fr_free(proofValuesRoot)
  let proofValuesX = ffi_rln_proof_values_get_x(proofValues)
  printFr("x", proofValuesX)
  ffi_fr_free(proofValuesX)
  let proofValuesExternalNullifier =
    ffi_rln_proof_values_get_external_nullifier(proofValues)
  printFr("external nullifier", proofValuesExternalNullifier)
  ffi_fr_free(proofValuesExternalNullifier)

  echo "\nVerifying proof"
  let verifyResult = ffi_rln_verify_with_roots(rlnInstance,
      rlnProof, addr roots, x)
  if verifyResult.err.dataPtr != nil:
    stderr.writeLine("Proof verification error: " & asString(verifyResult.err))
    ffi_c_string_free(verifyResult.err)
    return
  if verifyResult.ok:
    echo "  - proof verified successfully"
  else:
    echo "Proof verification failed"
    return

  ffi_rln_proof_values_free(proofValues)
  ffi_rln_proof_free(rlnProof)
  ffi_rln_witness_input_free(witness)
  ffi_rln_merkle_proof_free(merkleProof)
  ffi_fr_free(messageId)
  ffi_fr_free(x)
  ffi_fr_free(externalNullifier)
  ffi_vec_fr_free(roots)
  ffi_fr_free(computedRoot)
  ffi_vec_fr_free(pathElements)
  for i in 0 ..< treeDepth - 1:
    ffi_fr_free(defaultHashes[i])
  ffi_fr_free(defaultLeaf)
  memberFree(member)
  ffi_rln_free(rlnInstance)

main()
